### PR TITLE
feat: support config custom labels when scraping metrics

### PR DESCRIPTION
**Pull request by Xunzhuo**
Signed-off-by: Xunzhuo <mixdeers@gmail.com>

**Describe the change**

support config custom labels when scraping metrics in multi-cluster scenario.

prevent mixed topology when prometheus recieves more than one meshes metrics.

config like below:

``` yaml
  prometheus:
    url: "http://127.0.0.1:9090"
    custom_labels:
      mesh: "meshID"
      cluster: "clusterID"
      ...
```

config key and values of custom labels.

**Issue reference**

https://github.com/kiali/kiali/issues/4664

**Backwards incompatible?**

No

**Documentation**

* Does your contribution contain user-visible changes like e.g. new or changed configuration settings?

Yes, need some changes in CR reference.

**Before:**
**Namespace graph:**
image

**Node graph:**
image

**After:**
**Namespace graph:**
image
**Node graph:**
image

### CODE REVIEW

File: graph/telemetry/istio/istio.go [hunk 1]
```diff
@@ -800,6 +836,27 @@ func buildNodeTrafficMap(cluster, namespace string, n graph.Node, o graph.Teleme
...
 }
 
+// buildCustomLabelsFromConfig build custom addtional labels for prometheus from config
+func buildCustomLabelsFromConfig() string {
+	customLabels := config.Get().ExternalServices.Prometheus.CustomLabels
+	if len(customLabels) == 0 {
+		customLabels = config.Get().ExternalServices.CustomDashboards.Prometheus.CustomLabels
+		if len(customLabels) == 0 {
+			return ""
+		}
```
Comment (jmazzitelli) [inline graph/telemetry/istio/istio.go:846]: We don't want to do this. Here you are mixing the custom dashboards prometheus config with the regular prometheus config. The custom dashboards prometheus config is supposed to be isolated/separate from the "regular" prometheus config - the point being that the custom dashboards prometheus config points to a completely separate Prometheus instance than that of the Istio Prometheus. So one config should not assume the other - they are separate.

For details, see the kiali documentation:
* https://kiali.io/docs/configuration/custom-dashboard/#using-another-prometheus-instance
* https://kiali.io/docs/configuration/kialis.kiali.io/#.spec.external_services.custom_dashboards.prometheus

**Review by jmazzitelli — CHANGES_REQUESTED** [review]
Just gave this a quick glance, but found a flaw in the way the config is interpreted.

Two other things in addition to my code comment:

1. What does it mean to give custom labels to the custom dashboards prometheus config? Are these even used? If they are to be used, I don't see any new code in this PR that uses them. Code related to custom dashboard prometheus queries should be modified, too.
2. This is proposing new config to the Kiali CR which requires additional work over in the operator repo. Please see the documentation on what needs to be done: https://github.com/kiali/kiali-operator/blob/master/DEVELOPING.adoc#are-you-altering-a-kiali-server-configuration-setting

**Comment by jshaughn** [comment]
As much as I appreciate the submissions, your PR's are too premature.  This one isn't even close to being complete.  We don't really have the time to repeatedly review PRs that are very incrementally improved.  You need to better understand what you are doing before submitting PRs.  Please look fully at the code-base, and ask questions, before submitting a PR.  I don't think the discussion was even complete in the associated issue.  I had left it with a question about multi-tenancy.

**Review by jshaughn — CHANGES_REQUESTED** [review]
@Xunzhuo I think maybe I should generate PRs for this and you could use this as an exercise to review code as opposed to submit code, and in the process perhaps learn more about the various touch points.

To be clear, the idea you have here, and what is in this PR, I think is reasonable.

**Comment by jshaughn** [comment]
Hi @Xunzhuo , I think this original PR can be closed in favor of https://github.com/kiali/kiali/pull/4739.

**Comment by Xunzhuo** [comment]
> Hi @Xunzhuo , I think this original PR can be closed in favor of https://github.com/kiali/kiali/pull/4739.
> 
> 

thanks @jshaughn for helping me complete this PR, learned a lot!

**Comment by jshaughn** [comment]
No problem @Xunzhuo , thank you for your continued contributions to Kiali.  The next step in the process will be for QE to approve the https://github.com/kiali/kiali/pull/4739 PR, that will likely happen next week, and so I would expect this feature to be in Kiali 1.48.  Until then, please feel free to use/test the PRs.

**Pull request closed without merging**